odl/trafos/util/ft_utils.py

```diff
@@ -0,0 +1,654 @@
+        This option can only be used if ``space`` is a space of
+        real-valued functions.
+    shift : bool or `sequence` of bools, optional
+        If ``True``, the reciprocal grid is shifted by half a stride in
+        the negative direction. With a boolean sequence, this option
+        is applied separately to each axis.
+        If a sequence is provided, it must have the same length as
+        ``axes`` if supplied. Note that this must be set to ``True``
+        in the halved axis in half-complex transforms.
+        Default: ``True``
+    exponent : float, optional
+        Create a space with this exponent. By default, the conjugate
+        exponent ``q = p / (p - 1)`` of the exponent of ``space`` is
+        used, where ``q = inf`` for ``p = 1`` and vice versa.
+    dtype : optional
+        Complex data type of the reciprocal space. By default, the
+        complex counterpart of ``space.dtype`` is used.
+
+    Returns
+    -------
+    rspace : `DiscreteLp`
+        Reciprocal of the input ``space``. If ``halfcomplex=True``, the
+        upper end of the domain (where the half space ends) is chosen to
+        coincide with the grid node.
+    """
+    if not isinstance(space, DiscreteLp):
+        raise TypeError('`space` {!r} is not a `DiscreteLp` instance'
+                        ''.format(space))
+    if not space.is_uniform:
+        raise ValueError('`space` is not uniformly discretized')
+
+    if axes is None:
+        axes = list(range(space.ndim))
+    else:
+        try:
+            axes = [int(axes)]
+        except TypeError:
+            axes = list(axes)
+
+    if halfcomplex and space.field != RealNumbers():
+        raise ValueError('`halfcomplex` option can only be used with real '
+                         'spaces')
+
+    exponent = kwargs.pop('exponent', None)
+    if exponent is None:
+        exponent = conj_exponent(space.exponent)
+
+    dtype = kwargs.pop('dtype', None)
+    if dtype is None:
+        if is_real_dtype(space.dtype):
+            dtype = TYPE_MAP_R2C[space.dtype]
+        else:
+            dtype = space.dtype
+    else:
+        if not is_complex_floating_dtype(dtype):
+            raise ValueError('{} is not a complex data type'
+                             ''.format(dtype_repr(dtype)))
+
+    # Calculate range
+    recip_grid = reciprocal_grid(space.grid, shift=shift,
+                                 halfcomplex=halfcomplex, axes=axes)
+
+    # Make a partition with nodes on the boundary in the last transform axis
+    # if `halfcomplex == True`, otherwise a standard partition.
+    if halfcomplex:
+        max_pt = {axes[-1]: recip_grid.max_pt[axes[-1]]}
+        part = uniform_partition_fromgrid(recip_grid, max_pt=max_pt)
+    else:
+        part = uniform_partition_fromgrid(recip_grid)
+
+    recip_spc = uniform_discr_frompartition(part, exponent=exponent,
+                                            dtype=dtype, impl='numpy')
```

Comment:
We never check that the input has impl == 'numpy' so is it really fair to make the output have it? Shouldnt we accept any type of space anyway? At some point we could add CUFFT as a backend for cuda.